### PR TITLE
Update vertical xgb examples for gpu support

### PR DESCRIPTION
**Pull request by SYangster**
Updated for recently added vertical xgboost gpu support.

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [ ] Quick tests passed locally by running `./runtest.sh`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated.

### CODE REVIEW

File: examples/advanced/vertical_xgboost/README.md [hunk 1]
```diff
@@ -47,7 +47,13 @@ Next, we can use `FedXGBHistogramExecutor` and set XGBoost training parameters i
...
 Lastly, we must subclass `XGBDataLoader` and implement the `load_data()` method. For vertical federated learning, it is important when creating the `xgb.Dmatrix` to set `data_split_mode=1` for column mode, and to specify the presence of a label column `?format=csv&label_column=0` for the csv file. To support PSI, the dataloader can also read in the dataset based on the calculated intersection, and split the data into training and validation.
 
-> **_NOTE:_** For secure mode, make sure to provide the required certificates for the federated communicator. Also as of now, GPUs are not yet supported by vertical federated XGBoost.
+> **_NOTE:_** For secure mode, make sure to provide the required certificates for the federated communicator.
+
+### GPU Support
+By default, CPU based training is used.
+
+In order to enable GPU accelerated training, first ensure that your machine has CUDA installed and has at least two GPUs to simulate two clients.
+Then in `config_fed_client.json` set `"use_gpus": true` and  `"tree_method": "hist"` in `xgb_params`.
```
Comment (YuanTingHsieh) [inline examples/advanced/vertical_xgboost/README.md:56]: Do we really need 2 GPUs for simulating 2 clients?

Can we use 1 GPU to run 2 clients' training (parallel or one by one)?

File: nvflare/app_opt/xgboost/histogram_based/executor.py [hunk 2]
```diff
@@ -236,6 +246,10 @@ def train(self, shareable: Shareable, fl_ctx: FLContext, abort_signal: Signal) -
         self.rank = rank_map[client_name]
         self.world_size = world_size
 
+        if self.use_gpus:
+            self.log_info(fl_ctx, f"Training with GPU {self.rank}")
+            self.xgb_params["device"] = f"cuda:{self.rank}"
```
Comment (YuanTingHsieh) [inline nvflare/app_opt/xgboost/histogram_based/executor.py:251]: Let's add either a comment here or in the method docstring, that we are mapping each rank to a GPU.

As enhancement later (not sure if we need), we could say map RANK 0,1 to GPU 0, and RANK 2 to GPU 2.

File: examples/advanced/vertical_xgboost/README.md (hunk 1 above)
Comment (SYangster) [reply examples/advanced/vertical_xgboost/README.md:56]: yep, we need 2 GPUs with nccl, a single GPU will not work [https://github.com/dmlc/xgboost/pull/9552](ref)

**Comment by SYangster** [comment]
> Make sure you tested this in Multi-GPU

My machine only has 1 gpu, and I'm having issues building xgboost with nccl from source on our shared multi-gpu DGX due to outdated software, cuda etc. For this pr I was referencing the [changes](https://github.com/dmlc/xgboost/pull/9552) Rong made but I can mark as draft for now until we get the multi-gpu testing  working.

**Review by YuanTingHsieh — DISMISSED** [review]
Thanks!

Changes LGTM.

We need to rebase on main and make sure these examples work.

File: examples/advanced/vertical_xgboost/README.md [hunk 3]
```diff
@@ -47,7 +47,13 @@ Next, we can use `FedXGBHistogramExecutor` and set XGBoost training parameters i
...
 
-> **_NOTE:_** For secure mode, make sure to provide the required certificates for the federated communicator. Also as of now, GPUs are not yet supported by vertical federated XGBoost.
+> **_NOTE:_** For secure mode, make sure to provide the required certificates for the federated communicator.
+
+### GPU Support
+By default, CPU based training is used.
+
+In order to enable GPU accelerated training, first ensure that your machine has CUDA installed and has at least two GPUs to simulate two clients.
+In `config_fed_client.json` set `"use_gpus": true` and  `"tree_method": "hist"` in `xgb_params`. Then, in `FedXGBHistogramExecutor` we use the `device` parameter to map each rank to a GPU device ordinal in `xgb_params` (alternatively, you can simulate multiple GPUs with one only GPU by mapping each rank to the same GPU device).
```
Comment (YuanTingHsieh) [inline examples/advanced/vertical_xgboost/README.md:56]: Maybe we also need to correct the words here?
"has at least two GPUs" this is not a hard requirements anymore right?

Comment (SYangster) [reply examples/advanced/vertical_xgboost/README.md:56]: yes good catch

**Review by YuanTingHsieh — APPROVED** [review]
LGTM

**Pull request merged**